// @ngxs/store/esm2022/src/plugin-manager.mjs
import { inject, Injectable } from '@angular/core';
import { NGXS_PLUGINS } from '@ngxs/store/plugins';
import * as i0 from "@angular/core";
export class PluginManager {
    constructor() {
        this.plugins = [];
        this._parentManager = inject(PluginManager, {
            optional: true,
            skipSelf: true
        });
        this._pluginHandlers = inject(NGXS_PLUGINS, {
            optional: true
        });
        this.registerHandlers();
    }
    get _rootPlugins() {
        return this._parentManager?.plugins || this.plugins;
    }
    registerHandlers() {
        const pluginHandlers = this.getPluginHandlers();
        this._rootPlugins.push(...pluginHandlers);
    }
    getPluginHandlers() {
        const handlers = this._pluginHandlers || [];
        return handlers.map((plugin) => (plugin.handle ? plugin.handle.bind(plugin) : plugin));
    }
    /** @nocollapse */ static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.0.6", ngImport: i0, type: PluginManager, deps: [], target: i0.ɵɵFactoryTarget.Injectable }); }
    /** @nocollapse */ static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "18.0.6", ngImport: i0, type: PluginManager }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.0.6", ngImport: i0, type: PluginManager, decorators: [{
            type: Injectable
        }], ctorParameters: () => [] });
