// @ngxs/router-plugin/esm2022/src/router.module.mjs
import { NgModule, makeEnvironmentProviders } from '@angular/core';
import { NgxsModule, provideStates } from '@ngxs/store';
import { ɵcreateRouterPluginOptions, ɵNGXS_ROUTER_PLUGIN_OPTIONS, ɵUSER_OPTIONS } from '@ngxs/router-plugin/internals';
import { RouterState } from './router.state';
import { DefaultRouterStateSerializer, RouterStateSerializer } from './serializer';
import * as i0 from "@angular/core";
import * as i1 from "@ngxs/store";
export class NgxsRouterPluginModule {
    static forRoot(options) {
        return {
            ngModule: NgxsRouterPluginModule,
            providers: [
                { provide: ɵUSER_OPTIONS, useValue: options },
                {
                    provide: ɵNGXS_ROUTER_PLUGIN_OPTIONS,
                    useFactory: ɵcreateRouterPluginOptions,
                    deps: [ɵUSER_OPTIONS]
                },
                { provide: RouterStateSerializer, useClass: DefaultRouterStateSerializer }
            ]
        };
    }
    /** @nocollapse */ static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.0.6", ngImport: i0, type: NgxsRouterPluginModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    /** @nocollapse */ static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "18.0.6", ngImport: i0, type: NgxsRouterPluginModule, imports: [i1.ɵNgxsFeatureModule] }); }
    /** @nocollapse */ static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "18.0.6", ngImport: i0, type: NgxsRouterPluginModule, imports: [NgxsModule.forFeature([RouterState])] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.0.6", ngImport: i0, type: NgxsRouterPluginModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [NgxsModule.forFeature([RouterState])]
                }]
        }] });
export function withNgxsRouterPlugin(options) {
    return makeEnvironmentProviders([
        provideStates([RouterState]),
        { provide: ɵUSER_OPTIONS, useValue: options },
        {
            provide: ɵNGXS_ROUTER_PLUGIN_OPTIONS,
            useFactory: ɵcreateRouterPluginOptions,
            deps: [ɵUSER_OPTIONS]
        },
        { provide: RouterStateSerializer, useClass: DefaultRouterStateSerializer }
    ]);
}
