// @ngxs/router-plugin/esm2022/src/router.actions.mjs
/**
 * Public event api of the router
 */
export class Navigate {
    static { this.type = '[Router] Navigate'; }
    constructor(path, queryParams, extras) {
        this.path = path;
        this.queryParams = queryParams;
        this.extras = extras;
    }
}
/**
 *
 * Angular Routers internal state events
 *
 */
/**
 * An action dispatched when the router starts the navigation.
 */
export class RouterRequest {
    static { this.type = '[Router] RouterRequest'; }
    constructor(routerState, event, trigger = 'none') {
        this.routerState = routerState;
        this.event = event;
        this.trigger = trigger;
    }
}
/**
 * An action dispatched when the router navigates.
 */
export class RouterNavigation {
    static { this.type = '[Router] RouterNavigation'; }
    constructor(routerState, event, trigger = 'none') {
        this.routerState = routerState;
        this.event = event;
        this.trigger = trigger;
    }
}
/**
 * An action dispatched when the router cancel navigation.
 */
export class RouterCancel {
    static { this.type = '[Router] RouterCancel'; }
    constructor(routerState, storeState, event, trigger = 'none') {
        this.routerState = routerState;
        this.storeState = storeState;
        this.event = event;
        this.trigger = trigger;
    }
}
/**
 * An action dispatched when the router errors.
 */
export class RouterError {
    static { this.type = '[Router] RouterError'; }
    constructor(routerState, storeState, event, trigger = 'none') {
        this.routerState = routerState;
        this.storeState = storeState;
        this.event = event;
        this.trigger = trigger;
    }
}
/**
 * An action dispatched when the `ResolveEnd` event is triggered.
 */
export class RouterDataResolved {
    static { this.type = '[Router] RouterDataResolved'; }
    constructor(routerState, event, trigger = 'none') {
        this.routerState = routerState;
        this.event = event;
        this.trigger = trigger;
    }
}
/**
 * An action dispatched when the router navigation has been finished successfully.
 */
export class RouterNavigated {
    static { this.type = '[Router] RouterNavigated'; }
    constructor(routerState, event, trigger = 'none') {
        this.routerState = routerState;
        this.event = event;
        this.trigger = trigger;
    }
}
